src/include/migraphx/instruction.hpp

```diff
@@ -67,7 +74,10 @@ struct instruction
 
     static void backreference(instruction_ref ref);
 
-    static void replace_argument(instruction_ref ins, instruction_ref old, instruction_ref new_ins);
+    static void replace_argument(instruction_ref ins,
+                                 instruction_ref old,
+                                 instruction_ref new_ins,
+                                 bool recompute_shape = true);
```

Comment:
Remove this parameter, the shape should always be recomputed.

Reply:
Below shows call chain.  At the bottom of call chain,  the code recomputes shape for outputs.

When I concat nodes layer by layer,  each layer adjust its shape.  After the adjustment, it doesn't have the same dimension as its output layer. So this assertion fires:
 check_shapes{inputs}.has(2).same_type().same_dims();

That's why I need this parameter here, and also can't use "recompute_shape" to change shapes for instructions.




 instruction::replace_argument(instruction_ref ins,
                                   instruction_ref old,
                                   instruction_ref new_ins)
{
    ins->replace_argument(old, new_ins);
    backreference(ins);
    ins->recompute_shape();
}

void instruction::recompute_shape() { replace(compute_shape(op, arguments)); }

void instruction::replace(const shape& r)
{
    if(r != result)
    {
        result = r;
        for(auto&& ins : output)
        {
            assert(ins->name().front() != '@');
            ins->recompute_shape();
        }
    }
}

Reply:
Also, replacing an instruction's argument leads to re-computation of its output's shape doesn't look like a good design.   If you want to go down that path, how about output's output and so on?

Reply:
> When I concat nodes layer by layer, each layer adjust its shape. After the adjustment, it doesn't have the same dimension as its output layer. So this assertion fires:

Thats not a valid transformation.

The instruction class used to make all its fields public however, we ran into lots of issue where the fields would become inconsistent or wrong. Instead, we encapsulate them in these methods and ensure that every transformation with `replace_instruction` or `insert_instruction` is valid.

>  If you want to go down that path, how about output's output and so on?

Thats what `recompute_shape` already does.

Reply:
Your design does not take into account of the need of graph transformations. 
I can't change output chains during the transformation.

Reply:
You can;t demand a transformation to change all output chains in one call.  That's unreasonable. The design of optimization is a state-machine.  From one state to another state is an automated process.   You can't demand a transformation to skip all intermediate states.   So, unless you don't want to do any "complicated" work, you need to change the design to support it.

Reply:
I think its reasonable considering the issues we have dealt with in the past. And I dont see how this makes graph transformations impossible.

Reply:
A->B->C
when you change shape for A,  it triggers changing of shape for B.   But before B's shape is changed, you have an assertion that checks B's shape matches A's shape in certain dimensions.  How can this be logically correct?

Reply:
> How can this be logically correct?

I dont see how that could be correct, which is why we prevent such transformation.

However, if you want to change A, B, and C, then you should insert A', B' and C'(ie `A' -> B' -> C'`), and then replace C with C' and if thats a valid program there should be no exception thrown.

Reply:
It is unreasonable request.  You are saying we can't change graph in-place, we have to make copies and add changes to those copies.   In another words, we have to work around all possible means so that your design can't be changed.

Reply:
The fields are inconsistent or wrong is because you cache the field.  And these fields are derived from other primary inputs.  You shouldn't cache fields.

Reply:
I dont see how cache is the issue. If we have an instruction that changes its operator in a way that the output shape no longer is acceptable for its dependents, then we were not catching the issue until much later. Instead, we require every transformation to the graph to be a valid transformation so we can catch the problem much quicker.

I do not see how that is unreasonable. Its actually been quite helpful as we can track down incorrect graph transformations much quicker.

Reply:
As I mentioned in the meeting, one reason that graph is inconsistent is due to your design to cache fields.  And, with n->m, l->m, you can't compute m's shape when you are processing n.    Your idea to do x->y'->z assumes that y only has x as input.